cmd/remote-session.go

```diff
@@ -0,0 +1,222 @@
+package main
+
+import (
+	"fmt"
+	"github.com/spf13/cobra"
+	"os"
+	"os/exec"
+	"strings"
+)
+
+type RemoteSessionOptions struct {
+	Image      string
+	Expiration string
+}
+
+var (
+	remoteSessionOpts RemoteSessionOptions
+
+	cmdRemoteSession = &cobra.Command{
+		Use:   "remote-session",
+		Short: "cosa remote-session [command]",
+		Long:  "Initiate and use remote sessions for COSA execution.",
+	}
+
+	cmdRemoteSessionCreate = &cobra.Command{
+		Use:   "create",
+		Short: "Create a remote session",
+		Long: "Create a remote session. This command will print an ID to " +
+			"STDOUT that should be set in COREOS_ASSEMBLER_REMOTE_SESSION " +
+			"environment variable for later commands to use.",
+		RunE: runCreate,
+	}
+
+	cmdRemoteSessionDestroy = &cobra.Command{
+		Use:   "destroy",
+		Short: "Destroy a remote session",
+		Long: "Destroy a remote session. After running this command the " +
+			"COREOS_ASSEMBLER_REMOTE_SESSION should be unset.",
+		RunE: runDestroy,
+	}
+
+	cmdRemoteSessionExec = &cobra.Command{
+		Use:   "exec",
+		Short: "Execute a cosa command in the remote session",
+		Long:  "Execute a cosa command in the remote session.",
+		Args:  cobra.MinimumNArgs(1),
+		RunE:  runExec,
+	}
+
+	cmdRemoteSessionPS = &cobra.Command{
+		Use:   "ps",
+		Short: "Check if the remote session is running",
+		Long:  "Check if the remote session is running.",
+		Args:  cobra.ExactArgs(0),
+		RunE:  runPS,
+	}
+
+	cmdRemoteSessionSync = &cobra.Command{
+		Use:   "sync",
+		Short: "sync files/directories to/from the remote",
+		Long: "sync files/directories to/from the remote. The symantics here " +
+			"are similar to rsync or scp. Provide `:from to` or `from :to`. " +
+			"The argument with the leading ':' will represent the remote.",
+		Args: cobra.ExactArgs(2),
+		RunE: runSync,
+	}
+)
+
+func checkEnv(vars ...string) error {
+
+	// See if the user passed in any explicit vars to check. If
+	// not we'll go with the default set.
+	if len(vars) == 0 {
+		// We need to make sure that the CONTAINER_HOST and
+		// COREOS_ASSEMBLER_REMOTE_SESSION env vars are set
+		// We could check `CONTAINER_SSHKEY` key here but it's
+		// not strictly required (user could be using ssh-agent).
+		vars = []string{"CONTAINER_HOST", "COREOS_ASSEMBLER_REMOTE_SESSION"}
+	}
+	for _, v := range vars {
+		val, ok := os.LookupEnv(v)
+		if !ok || val == "" {
+			return fmt.Errorf("The env var %s must be defined and non-empty", v)
+		}
+	}
+	return nil
+}
+
+// Creates a "remote session" on the remote. This just creates a
+// container on the remote and prints to STDOUT the container ID.
+// The user is then expected to store this ID in the
+// COREOS_ASSEMBLER_REMOTE_SESSION environment variable.
+func runCreate(c *cobra.Command, args []string) error {
+	if err := checkEnv("CONTAINER_HOST"); err != nil {
```

Comment:
Hmm, all the other calls require both, but this one only requires `CONTAINER_HOST`. I think it'd be good if we also checked for `COREOS_ASSEMBLER_REMOTE_SESSION` and errored out if it's defined.

So maybe a cleaner API is to change `checkEnv()` to just take a `bool` arg like `expect_session` and pass it as `false` everywhere else except here.

Reply:
before I saw this comment I had reworked `checkEnv()` into `preRunCheckEnv()` (see latest upload).

That new update still doesn't address what you want, though. Which is to error out if we create and `COREOS_ASSEMBLER_REMOTE_SESSION` is already set.

Reply:
Ok should be fixed now.